source/common/router/router.cc

```diff
@@ -1130,12 +1151,33 @@ void Filter::UpstreamRequest::encodeTrailers(const Http::HeaderMap& trailers) {
   if (!request_encoder_) {
     ENVOY_STREAM_LOG(trace, "buffering trailers", *parent_.callbacks_);
   } else {
+    // Encodes metadata if exists.
+    if (!parent_.downstream_metadata_map_vector_.empty()) {
```

Comment:
Oops - I'd missed that we were using the same map location for buffered metadata. 
I think we'll want this to be per-upstream for when hedging (#6228) lands.

At that point we'll create a copy of metadata for each  upstream and each upstream will keep its own queue.

Reply:
Good catch! Moved metadata storage to upstream.